rust/src/openssl.rs

```diff
@@ -0,0 +1,82 @@
+//! [PrivateKey] and [PublicKey] implementations for types in [openssl].
+//!
+//! Although X25519 is specified to never need exception handling, the openssl
+//! crate returns [Result] values. The implementations in this module panic on
+//! returned errors.
+//!
+//! Usage example:
+//!
+//! ```
+//! use openssl::pkey;
+//! use glome::{tag,verify};
+//! let raw_private_key = [15u8; 32]; // <- our private key loaded from somewhere
+//! let private_key = pkey::PKey::private_key_from_raw_bytes(&raw_private_key, pkey::Id::X25519).unwrap();
+//! let raw_public_key = [16u8; 32]; // <- their public key loaded from somewhere
+//! let public_key = pkey::PKey::public_key_from_raw_bytes(&raw_public_key, pkey::Id::X25519).unwrap();
+//! tag(&private_key, &public_key, 0u8, b"Hello, world!");
+//! verify(&private_key, &public_key, 42u8, b"Goodbye.", b"\x00\x01\x02\x03");
+//! ```
+
+use openssl::{derive, pkey};
+
+use crate::{PrivateKey, PublicKey};
+
+impl PublicKey for pkey::PKey<pkey::Public> {
+    fn to_bytes(&self) -> [u8; 32] {
+        self.raw_public_key()
+            .expect("an X25519 key should be convertible to bytes")
+            .try_into()
+            .expect("raw public key should be 32 bytes long")
+    }
+}
+
+impl PrivateKey for pkey::PKey<pkey::Private> {
+    type PublicKey = pkey::PKey<pkey::Public>;
+
+    fn dh(&self, theirs: &Self::PublicKey) -> [u8; 32] {
+        let mut deriver = derive::Deriver::new(self)
+            .expect("should be able to create a deriver from an X25519 key");
+        deriver
+            .set_peer(theirs)
+            .expect("should be able to set an X25519 public key as peer");
+
+        let mut secret = [0u8; 32];
+        let n = deriver
+            .derive(&mut secret)
+            .expect("should be able to derive shared secret");
+        assert_eq!(n, 32); // TODO(burgerdev): can I use assert_eq here?
```

Comment:
Yes! You can use assert everywhere. It will just panic if the check fails and I guess this is exactly what you want here.

Reply:
Thanks!